### PR TITLE
Add support for nested metrics in CompileMetrics (fixes #21700)

### LINKED ISSUE #21700: Nested outputs structures and corresponding nested losses are supported, but not nested metrics
This is a follow up to https://github.com/keras-team/keras/issues/21675 and https://github.com/keras-team/keras/pull/21694

- nested output structures are supported
- nested losses matching the nested output structures are supported
- however, the same is not true for metrics, the code within the `try`/`except` block fails

```python
import numpy as np
import keras

print('keras version:', keras.__version__)

X = np.random.rand(100, 32)
Y1 = np.random.rand(100, 10)
Y2 = np.random.rand(100, 10)
Y3 = np.random.rand(100, 10)

def create_model():
    x = keras.Input(shape=(32,))
    y1 = keras.layers.Dense(10)(x)
    y2 = keras.layers.Dense(10)(x)
    y3 = keras.layers.Dense(10)(x)
    return keras.Model(inputs = x, outputs = {'a': y1, 'b': {'c': y2, 'd': y3}})

model = create_model()
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}})
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training not using metrics\n')

try:
    model = create_model()
    model.compile(
        optimizer = 'adam',
        loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
        metrics = {'a': ['mae', 'acc'], 'b':{'c': 'mse', 'd': 'mse'}}
    )
    model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
    print('Successful training using metrics')
except ValueError as ex:
    print('Unsuccessful training with metrics')
    print(ex)

model = create_model()
model.compile(
    optimizer = 'adam',
    loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
    loss_weights = {'a': 1., 'b': {'c': 0.5, 'd': 0.4}}
)
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training using loss_weights\n')
```

Ideally, metrics and losses should be handled the same. Part of the complication is that there is a mechanism to use the output names when not using the output structure.

Another complication is that metrics can be either a single metric or a list of metrics for each output. The logic that turns single metrics to a list must handle nested structures correctly (with `keras.tree.map_structure_up_to`).

### PR DESCRIPTION
**Pull request by yashwantbezawada**
## Summary
- Adds support for nested dict metrics in `CompileMetrics` to match nested output structures
- Follows the existing `CompileLoss._build_nested()` pattern for recursive structure handling
- Fixes #21700

## Changes
- Modified `MetricsList` class to use `output_path` tuple for path-based tracking
- Added `resolve_path()` helper function to navigate nested structures
- Added `_is_nested_config()` method to detect nested dict configurations
- Added `_build_nested_metrics()` recursive method following `CompileLoss` pattern
- Updated `build()` and `update_state()` methods for path-based resolution
- Added 6 comprehensive test cases for nested metrics functionality

## Test plan
- [x] All existing `compile_utils_test.py` tests pass (backward compatibility)
- [x] New test cases cover: basic nested dict, deeply nested (3+ levels), mixed nested/flat, weighted metrics with nesting, and error handling for invalid keys
- [x] Code follows Keras contribution guidelines (80 char line limit, Ruff formatting)

### CODE REVIEW

File: keras/src/trainers/compile_utils_test.py [hunk 1]
```diff
@@ -286,6 +286,181 @@ def test_dict_outputs_ignore_mismatched_output_names(self):
...
+            "a": np.array([1.0, 2.0]),
+            "b": {"c": np.array([0.5, 1.5])},
+        }
+
+        compile_metrics.build(y_true, y_pred)
+        compile_metrics.update_state(y_true, y_pred, sample_weight)
+        result = compile_metrics.result()
+
+        self.assertIn("a_mse", result)
+        self.assertIn("b_c_mae", result)
```
[the remark this answers is not in the document]
Comment (yashwantbezawada) [reply keras/src/trainers/compile_utils_test.py:423]: Done. Added value assertions.

File: keras/src/trainers/compile_utils.py [hunk 2]
```diff
@@ -174,6 +196,149 @@ def variables(self):
...
+                zip(metrics, y_true, y_pred)
+            ):
+                self._build_nested_metrics(
+                    sub_metrics,
+                    sub_yt,
+                    sub_yp,
+                    flat_list,
+                    current_path + (idx,),
+                    argument_name,
+                )
```
[the remark this answers is not in the document]
Comment (yashwantbezawada) [reply keras/src/trainers/compile_utils.py:332]: Done. Extracted `_resolve_leaf_tensors` helper.

File: keras/src/trainers/compile_utils.py [hunk 3]
```diff
@@ -336,23 +530,83 @@ def _flatten_y(self, y):
...
+            else:
+                sample_weight_flat = [None for _ in range(len(y_true_flat))]
+            for m, y_t, y_p, s_w in zip(
+                self._flat_weighted_metrics,
+                y_true_flat,
+                y_pred_flat,
+                sample_weight_flat,
+            ):
+                if m:
+                    m.update_state(y_t, y_p, s_w)
```
[the remark this answers is not in the document]
Comment (yashwantbezawada) [reply keras/src/trainers/compile_utils.py:601]: The two loops differ - one passes sample_weight, one doesn't. Keeping inline is clearer.

**Comment by yashwantbezawada** [comment]
Hey, just checking if this is on the radar for review. Happy to address any feedback.

**Comment by hertschuh** [comment]
Closing as this was addressed in https://github.com/keras-team/keras/pull/22308 which also fixes some other issues and reduces the amount of code.
